File: scripts/docs/examples/closed-parts.ts
/** The closed parts of the scenes built of thin solids (the chalet, #415): boxes and octagonal
 *  logs, each a closed solid whose faces point away from its centre, with the normals an exporter
 *  writes. The compiler's arcade test builds the same parts in Rust (`closed_parts.rs`): keep the
 *  two in step. */
import type { Mesh } from './mesh.ts';
import type { Vec3 } from './random.ts';

/** A log's radius and the segments along it. */
export const LOG_RADIUS = 0.12;
const LOG_SEGMENTS = 8;

/** One corner, its position and its normal; three make a triangle. */
export type Corner = readonly [Vec3, Vec3];
export type Triangle = readonly [Corner, Corner, Corner];

/** The point whose coordinate on axis `a` is `coordinate(a)`. */
export const point = (coordinate: (a: number) => number): Vec3 =>
  [0, 1, 2].map(coordinate) as never;
const minus = (a: Vec3, b: Vec3) => point((k) => a[k] - b[k]);
/** The point or direction with `values` on axes `axes`, zero elsewhere. */
const on = (axes: readonly number[], values: readonly number[]) =>
  point((a) => values[axes.indexOf(a)] ?? 0);

/**
 * Pushes triangles given in pairs, each pair a quad, turned to face away from their convex part's
 * `centre`. Corners of one quad equal in position and normal are one vertex; each quad writes its
 * own, as the open world's exporter writes them: twins in everything a page stores.
 */
export function pushPart(mesh: Mesh, triangles: readonly Triangle[], centre: Vec3) {
  const shared = new Map<string, number>();
  triangles.forEach((triangle, rank) => {
    const [a, b, c] = triangle.map(([p]) => p),
      [u, v, out] = [minus(b, a), minus(c, a), minus(a, centre)];
    const facing =
      (u[1] * v[2] - u[2] * v[1]) * out[0] +
      (u[2] * v[0] - u[0] * v[2]) * out[1] +
      (u[0] * v[1] - u[1] * v[0]) * out[2];
    for (const k of facing < 0 ? [0, 2, 1] : [0, 1, 2]) {
      const [position, normal] = triangle[k],
        key = `${rank >> 1} ${position} ${normal}`;
      let index = shared.get(key);
      if (index === undefined) {
        shared.set(key, (index = mesh.positions.length / 3));
        mesh.positions.push(...position);
        mesh.normals.push(...normal);
      }
      mesh.indices.push(index);
    }
  });
}

/** The twelve triangles of a closed box, flat faces, unordered. */
export function boxTriangles(min: Vec3, max: Vec3): Triangle[] {
  const triangles: Triangle[] = [];
  for (let axis = 0; axis < 3; axis++) {
    const plane = [axis, (axis + 1) % 3, (axis + 2) % 3];
    for (const [side, sign] of [
      [min[axis], -1],
      [max[axis], 1],
    ]) {
      const corner = (u: number, v: number): Corner => [
        on(plane, [side, u, v]),
        on([axis], [sign]),
      ];
      const [a, b, c, d] = [
        corner(min[plane[1]], min[plane[2]]),
        corner(max[plane[1]], min[plane[2]]),
        corner(max[plane[1]], max[plane[2]]),
        corner(min[plane[1]], max[plane[2]]),
      ];
      triangles.push([a, b, c], [a, c, d]);
    }
  }
  return triangles;
}

/** A closed box. */
export function pushBox(mesh: Mesh, min: Vec3, max: Vec3) {
  pushPart(
    mesh,
    boxTriangles(min, max),
    point((a) => (min[a] + max[a]) / 2),
  );
}

/** A closed octagonal log, 0.12 m in radius, along `axis` from `start` over `length`, smooth sides and flat caps. */
export function pushLog(mesh: Mesh, axis: number, start: Vec3, length: number) {
  const s = Math.SQRT1_2,
    ring = [
      [1, 0],
      [s, s],
      [0, 1],
      [-s, s],
      [-1, 0],
      [-s, -s],
      [0, -1],
      [s, -s],
    ] as const,
    plane = [(axis + 1) % 3, (axis + 2) % 3];
  const at = (along: number, round: readonly number[]) =>
    point((a) => start[a] + (a === axis ? along : round[plane.indexOf(a)] * LOG_RADIUS));
  const side = (along: number, r: readonly number[]): Corner => [at(along, r), on(plane, r)];
  const triangles: Triangle[] = [],
    step = length / LOG_SEGMENTS;
  for (let segment = 0; segment < LOG_SEGMENTS; segment++)
    ring.forEach((r0, k) => {
      const r1 = ring[(k + 1) % ring.length],
        [a0, a1] = [segment * step, (segment + 1) * step];
      const q = [side(a0, r0), side(a1, r0), side(a1, r1), side(a0, r1)];
      triangles.push([q[0], q[1], q[2]], [q[0], q[2], q[3]]);
    });
  for (const [along, sign] of [
    [0, -1],
    [length, 1],
  ]) {
    const cap = (r: readonly number[]): Corner => [at(along, r), on([axis], [sign])];
    for (let k = 1; k < ring.length - 1; k++)
      triangles.push([cap(ring[0]), cap(ring[k]), cap(ring[k + 1])]);
  }
  pushPart(
    mesh,
    triangles,
    point((a) => start[a] + (a === axis ? length / 2 : 0)),
  );
}

File: packages/asset-compiler-rust/src/tests/compile/closed_parts.rs
//! Closed solids built in code for the compiler's tests (#484): boxes and octagonal 0.12 m logs,
//! each a closed part whose faces point away from its centre, with the normals an exporter writes,
//! and the one-node fixture that cooks them. The chalet of #415 is a committed scene instead,
//! built of the same parts by `scripts/docs/examples/closed-parts.ts`: keep the two in step.
use super::silhouette::Mesh;
use super::*;

/// The logs: 0.12 m in radius, octagonal, in eight segments.
const LOG_RADIUS: f32 = 0.12;
const LOG_SEGMENTS: usize = 8;

/// One triangle: per corner, its position and its normal.
type Triangle = [([f32; 3], [f32; 3]); 3];

/// Triangles pushed in pairs, each pair a quad, then turned to face away from their convex part's
/// centre. Corners of one quad equal in position and normal are one vertex; each quad writes its
/// own, as the open world's exporter writes them: twins in everything a page stores.
fn push_part(mesh: &mut Mesh, triangles: &[Triangle], centre: [f32; 3]) {
    let mut shared = std::collections::HashMap::new();
    for (rank, triangle) in triangles.iter().enumerate() {
        let p = triangle.map(|(p, _)| p.map(f64::from));
        let normal = crate::shared_math::cross(
            crate::shared_math::sub(p[1], p[0]),
            crate::shared_math::sub(p[2], p[0]),
        );
        let outward = crate::shared_math::sub(p[0], centre.map(f64::from));
        let order = if crate::shared_math::dot(normal, outward) < 0.0 {
            [0, 2, 1]
        } else {
            [0, 1, 2]
        };
        for k in order {
            let (position, normal) = triangle[k];
            let key = (
                rank / 2,
                position.map(f32::to_bits),
                normal.map(f32::to_bits),
            );
            let index = *shared.entry(key).or_insert_with(|| {
                mesh.positions.push(position.map(f64::from));
                mesh.normals.push(normal.map(f64::from));
                mesh.positions.len() as u32 - 1
            });
            mesh.indices.push(index);
        }
    }
}

/// A closed box, flat faces.
pub(super) fn push_box(mesh: &mut Mesh, min: [f32; 3], max: [f32; 3]) {
    let centre = std::array::from_fn(|a| (min[a] + max[a]) / 2.0);
    push_part(mesh, &box_triangles(min, max), centre);
}

/// The twelve triangles of a closed box, unordered.
fn box_triangles(min: [f32; 3], max: [f32; 3]) -> Vec<Triangle> {
    let mut triangles = Vec::new();
    for axis in 0..3 {
        let (u, v) = ((axis + 1) % 3, (axis + 2) % 3);
        for (side, sign) in [(min[axis], -1.0), (max[axis], 1.0)] {
            let mut normal = [0.0; 3];
            normal[axis] = sign;
            let corner = |a: f32, b: f32| {
                let mut p = [0.0; 3];
                (p[axis], p[u], p[v]) = (side, a, b);
                (p, normal)
            };
            let (a, b, c, d) = (
                corner(min[u], min[v]),
                corner(max[u], min[v]),
                corner(max[u], max[v]),
                corner(min[u], max[v]),
            );
            triangles.extend([[a, b, c], [a, c, d]]);
        }
    }
    triangles
}

/// A closed octagonal log along `axis` from `start` over `length`, smooth sides and flat caps.
pub(super) fn push_log(mesh: &mut Mesh, axis: usize, start: [f32; 3], length: f32) {
    let s = 0.5f32.sqrt();
    let ring = [
        (1.0, 0.0),
        (s, s),
        (0.0, 1.0),
        (-s, s),
        (-1.0, 0.0),
        (-s, -s),
        (0.0, -1.0),
        (s, -s),
    ];
    let (u, v) = ((axis + 1) % 3, (axis + 2) % 3);
    let point = |along: f32, (cu, cv): (f32, f32)| {
        let mut p = start;
        p[axis] += along;
        p[u] += cu * LOG_RADIUS;
        p[v] += cv * LOG_RADIUS;
        let mut n = [0.0; 3];
        (n[u], n[v]) = (cu, cv);
        (p, n)
    };
    let mut triangles = Vec::new();
    let step = length / LOG_SEGMENTS as f32;
    for segment in 0..LOG_SEGMENTS {
        let (a0, a1) = (segment as f32 * step, (segment + 1) as f32 * step);
        for k in 0..ring.len() {
            let (r0, r1) = (ring[k], ring[(k + 1) % ring.len()]);
            let q = [point(a0, r0), point(a1, r0), point(a1, r1), point(a0, r1)];
            triangles.extend([[q[0], q[1], q[2]], [q[0], q[2], q[3]]]);
        }
    }
    for (along, sign) in [(0.0, -1.0), (length, 1.0)] {
        let cap = |r| {
            let (p, _) = point(along, r);
            let mut n = [0.0; 3];
            n[axis] = sign;
            (p, n)
        };
        for k in 1..ring.len() - 1 {
            triangles.push([cap(ring[0]), cap(ring[k]), cap(ring[k + 1])]);
        }
    }
    let mut centre = start;
    centre[axis] += length / 2.0;
    push_part(mesh, &triangles, centre);
}

/// One node drawing `mesh` as one primitive, positions and normals as `f32`.
pub(super) fn mesh_fixture(tag: &str, mesh: &Mesh) -> (PathBuf, Options) {
    let mut buffer = GltfBuffer::default();
    let vec3 = |points: &[[f64; 3]]| -> Vec<u8> {
        points
            .iter()
            .flatten()
            .flat_map(|&v| (v as f32).to_le_bytes())
            .collect()
    };
    let count = mesh.positions.len();
    let (min, max) = (0..3).fold((vec![], vec![]), |(mut lo, mut hi), a| {
        let values = mesh.positions.iter().map(|p| p[a] as f32);
        lo.push(values.clone().fold(f32::INFINITY, f32::min));
        hi.push(values.fold(f32::NEG_INFINITY, f32::max));
        (lo, hi)
    });
    let position = buffer.push(
        vec3(&mesh.positions),
        json!({"componentType":5126,"type":"VEC3","count":count,"min":min,"max":max}),
    );
    let normal = buffer.push(
        vec3(&mesh.normals),
        json!({"componentType":5126,"type":"VEC3","count":count}),
    );
    let indices = mesh.indices.iter().flat_map(|v| v.to_le_bytes()).collect();
    let index = buffer.push(
        indices,
        json!({"componentType":5125,"type":"SCALAR","count":mesh.indices.len()}),
    );
    let primitive =
        json!({"attributes":{"POSITION":position,"NORMAL":normal},"indices":index,"material":0});
    let gltf = json!({"asset":{"version":"2.0"},"buffers":[{"uri":format!("{tag}.bin"),"byteLength":buffer.bin.len()}],
        "bufferViews":buffer.views,"accessors":buffer.accessors,"meshes":[{"primitives":[primitive]}],"nodes":[{"mesh":0}],
        "scenes":[{"nodes":[0]}],"scene":0,"materials":[{"pbrMetallicRoughness":{"metallicFactor":0.0}}]});
    gltf_fixture(tag, &gltf, &buffer.bin)
}
